Code/Framework/AzToolsFramework/AzToolsFramework/UI/Prefab/Procedural/ProceduralPrefabReadOnlyHandler.cpp

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <AzToolsFramework/UI/Prefab/Procedural/ProceduralPrefabReadOnlyHandler.h>
+
+#include <AzToolsFramework/Entity/EditorEntityContextBus.h>
+#include <AzToolsFramework/Entity/ReadOnly/ReadOnlyEntityInterface.h>
+#include <AzToolsFramework/Prefab/PrefabPublicInterface.h>
+#include <AzToolsFramework/Prefab/PrefabFocusPublicInterface.h>
+
+namespace AzToolsFramework
+{
+    namespace Prefab
+    {
+        ProceduralPrefabReadOnlyHandler::ProceduralPrefabReadOnlyHandler()
+        {
+            m_prefabPublicInterface = AZ::Interface<PrefabPublicInterface>::Get();
+            AZ_Assert(
+                m_prefabPublicInterface != nullptr,
+                "ProceduralPrefabReadOnlyHandler requires a PrefabPublicInterface instance on Initialize.");
+
+            m_prefabFocusPublicInterface = AZ::Interface<PrefabFocusPublicInterface>::Get();
+            AZ_Assert(
+                m_prefabFocusPublicInterface != nullptr,
+                "ProceduralPrefabReadOnlyHandler requires a PrefabFocusPublicInterface instance on Initialize.");
+
+            AzFramework::EntityContextId editorEntityContextId = AzFramework::EntityContextId::CreateNull();
+            EditorEntityContextRequestBus::BroadcastResult(editorEntityContextId, &EditorEntityContextRequests::GetEditorEntityContextId);
+
+            ReadOnlyEntityQueryRequestBus::Handler::BusConnect(editorEntityContextId);
+
+             // Refresh the whole read-only cache
+            if (auto readOnlyEntityQueryInterface = AZ::Interface<ReadOnlyEntityQueryInterface>::Get())
+            {
+                readOnlyEntityQueryInterface->RefreshReadOnlyStateForAllEntities();
+            }
+        }
+
+        ProceduralPrefabReadOnlyHandler ::~ProceduralPrefabReadOnlyHandler()
+        {
+            ReadOnlyEntityQueryRequestBus::Handler::BusDisconnect();
+        }
+
+        void ProceduralPrefabReadOnlyHandler::IsReadOnly(const AZ::EntityId& entityId, bool& isReadOnly)
+        {
+            if(m_prefabPublicInterface->IsOwnedByProceduralPrefabInstance(entityId))
+            {
+                // All entities nested inside a procedural prefabs should always be marked as read-only.
+                if (!m_prefabPublicInterface->IsInstanceContainerEntity(entityId))
+                {
+                    isReadOnly = true;
+                }
+
+                // The container entity of a procedural prefab should only be marked as read-only when the prefab is being edited.
+                if (m_prefabFocusPublicInterface->IsOwningPrefabBeingFocused(entityId))
+                {
+                    isReadOnly = true;
+                }
```

Comment:
suggestion/nit: There's likely a terser way to write this where we just assign to `isReadOnly` (but care must be taken not to change the behavior). It could be something like...

```
isReadOnly = !m_prefabPublicInterface->IsInstanceContainerEntity(entityId))
isReadOnly = isReadOnly || m_prefabFocusPublicInterface->IsOwningPrefabBeingFocused(entityId);
```

Also in this case having `IsReadOnly` return true/false might be a nicer API to use (unless an `in/out` parameter is strictly required).

```
return !m_prefabPublicInterface->IsInstanceContainerEntity(entityId) ||
    m_prefabFocusPublicInterface->IsOwningPrefabBeingFocused(entityId);
```

You probably want to write a test first just in case there's a subtle bug with the transformation 😛

Reply:
Since there could be multiple handlers for this function, we don't want to assign false as that could possibly override the result from other handlers.

Potentially I could turn this into
```
isReadOnly &= !m_prefabPublicInterface->IsInstanceContainerEntity(entityId));
isReadOnly &= isReadOnly || m_prefabFocusPublicInterface->IsOwningPrefabBeingFocused(entityId);
```

But I think the current code is more readable.

Reply:
Make sure not to use bitwise operators with boolean values, the version I posted above could be updated to

```
isReadOnly = !m_prefabPublicInterface->IsInstanceContainerEntity(entityId)) && isReadOnly;
isReadOnly = isReadOnly || m_prefabFocusPublicInterface->IsOwningPrefabBeingFocused(entityId);
```

Using bitwise operators with boolean values can have subtle bugs (e.g. issues with not short-circuiting and with truthy values that fail). I'll send you a link to a talk I gave on it a while back 😛

Reply:
Noted! I did not use the code above, but I'll keep that in mind :)